### PR TITLE
WebPush feature

### LINKED ISSUE #335: Having a "WebPush" device --does it make sense?
I started to use this project before fully understanding what is a WebPush and what is a Push. Given that I wanted to perform WebPush for Android Chrome browsers, I was focused on the GCM stuff and forgot to teach myself about WebPush and its requirements --mainly the required encryption mechanisms.

The current GCM device here is somewhat compatible with Chrome WebPush browsers, but because they lack encryption it is not possible to send information from the server to the client (at least, that is what I understood from [google](https://developers.google.com/web/updates/2016/03/web-push-encryption) and [mozilla](https://developer.mozilla.org/en-US/docs/Web/API/Push_API/Using_the_Push_API), sorry for my ignorance).

In the end I rolled my own library (my requirements were very modest) using the [`pywebpush`](https://github.com/web-push-libs/pywebpush/) which provides the encryption requirements for the WebPush I wanted. However, I was thinking that maybe a WebPushDevice would be a "good feature" to `django-push-notifications`.

Does it make sense? Or it would clutter the library and it is not the same philosophy as the current devices?

### PR DESCRIPTION
**Pull request by goinnn**
WebPush support for Chrome 50+, Firefox 50+ and Opera 47+

This is the main commit [7e567b7]. Next commits are: Opera support, and documentation.

Fix #335

### CODE REVIEW

File: push_notifications/conf/app.py [hunk 1]
```diff
@@ -282,3 +302,15 @@ def get_wns_package_security_id(self, application_id=None):
 
 	def get_wns_secret_key(self, application_id=None):
 		return self._get_application_settings(application_id, "WNS", "SECRET_KEY")
+
+	def get_wns_secret_key(self, application_id=None):
```
Comment (matthewh) [inline push_notifications/conf/app.py]: Did get_wns_secret_key get duplicated?

File: push_notifications/conf/app.py [hunk 2]
```diff
@@ -26,7 +26,8 @@
 	"APNS",
 	"FCM",
 	"GCM",
-	"WNS"
+	"WNS",
+	"BROWSER"
```
Comment (matthewh) [inline push_notifications/conf/app.py]: Why use the key **BROWSER** instead of **WEB** or **WEBP**?

Comment (jamaalscarlett) [reply push_notifications/conf/app.py]: What about WPNS (Web push notification service)?

File: setup.cfg [hunk 3]
```diff
@@ -1,7 +1,7 @@
 [metadata]
 name = django-push-notifications
-version = 1.5.0
-description = Send push notifications to mobile devices through GCM, APNS or WNS in Django
+version = 1.6.0
+description = Send push notifications to mobile devices through GCM, APNS, WNS or WebPush in Django
```
Comment (jamaalscarlett) [inline setup.cfg]: This should be updated if notifications can also be sent to browsers.

File: push_notifications/conf/app.py (hunk 2 above)
Comment (matthewh) [reply push_notifications/conf/app.py]: To better clarify my position. I think BROWSER is a little to generic. You can push to a browser (Google Chrome) using just FCM/GCM as well. @jamaalscarlett's suggestion is also good. Just not **BROWSER**.

File: README.rst [hunk 4]
```diff
@@ -93,6 +96,204 @@ For WNS, you need both the ``WNS_PACKAGE_SECURITY_KEY`` and the ``WNS_SECRET_KEY
...
 
+**Browser settings**
+
+Install:
+********
+
+.. code-block:: python
+
+	pip install pywebpush
+	pip install py-vapid
```
Comment (jamaalscarlett) [inline README.rst]: You might want to note that py-vapid is only required to generate the keys.  That does not necessarily have to occur on the server.

File: README.rst [hunk 5]
```diff
@@ -93,6 +96,204 @@ For WNS, you need both the ``WNS_PACKAGE_SECURITY_KEY`` and the ``WNS_SECRET_KEY
 - ``USER_MODEL``: Your user model of choice. Eg. ``myapp.User``. Defaults to ``settings.AUTH_USER_MODEL``.
 - ``UPDATE_ON_DUPLICATE_REG_ID``: Transform create of an existing Device (based on registration id) into a update. See below `Update of device with duplicate registration ID`_ for more details.
 
+**Browser settings**
+
+Install:
```
Comment (jamaalscarlett) [inline README.rst]: It's a nitpick, but the sub-headings should be tabbed over so it is clear that they are for the Webpush feature.

File: push_notifications/conf/app.py (hunk 1 above)
Comment (goinnn) [reply push_notifications/conf/app.py]: Yes of course, it is a silly error.

File: push_notifications/conf/app.py (hunk 2 above)
Comment (goinnn) [reply push_notifications/conf/app.py]: WPNS is very similar to WNS. Also the protocol name is WebPush. If you want I or you can change it for WP, but WP is Windows Phone acronym. 

I had doubts about this name

File: setup.cfg (hunk 3 above)
Comment (goinnn) [reply setup.cfg]: Yes of course

File: README.rst (hunk 4 above)
Comment (goinnn) [reply README.rst]: Yes, I or you can add a comment

File: README.rst (hunk 5 above)
Comment (goinnn) [reply README.rst]: Ok

**Comment by goinnn** [comment]
@jamaalscarlett @matthewh I have fixed the duplicated error code.

@jamaalscarlett @matthewh Please tell me the name for "BROWSER" settings and I can change it. For me I think the better is "WEBPUSH".

@jamaalscarlett Please could you change the documentation. I am sure you can change it better than me. I am very strict with details too. And it is very difficult another people can write the description or documentation like I want it.

If you want I can share a preproduction site with an example.

File: push_notifications/conf/app.py (hunk 2 above)
Comment (jamaalscarlett) [reply push_notifications/conf/app.py]: I'm fine with WP

Comment (matthewh) [reply push_notifications/conf/app.py]: Go with **WP**. Windows phone has been EOLd so I do not think that will be a concern.

File: setup.cfg [hunk 6]
```diff
@@ -1,6 +1,6 @@
 [metadata]
 name = django-push-notifications
-version = 1.5.0
+version = 1.6.0
```
Comment (jleclanche) [inline setup.cfg]: Please do not bump the version, this is done during releases.

File: push_notifications/webpush.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		}
+	}
+
+
+def webpush_send_message(uri, message, browser, auth, p256dh, application_id=None, **kwargs):
+	try:
+		from pywebpush import webpush
+		from pywebpush import WebPushException
+	except ImportError:
+		raise WebPushError("Please install pywebpush package: pip install pywebpush")
```
Comment (jleclanche) [inline push_notifications/webpush.py]: This should not be a WebPushError. In fact, you can let the ImportError happen. If you'd like for this to be checked for, you can implement a django check that looks at your settings and checks the correct lib is present.

File: push_notifications/models.py [hunk 8]
```diff
@@ -206,3 +209,44 @@ def send_message(self, message, **kwargs):
...
+		for device in devices:
+			res.append(device.send_message(message))
+		return res
+
+
+class WebPushDevice(Device):
+	registration_id = models.TextField(verbose_name=_("Registration ID"))
+	p256dh = models.TextField(verbose_name=_("User public encryption key"))
+	auth = models.TextField(verbose_name=_("User auth secret"))
+	browser = models.CharField(
```
Comment (jleclanche) [inline push_notifications/models.py]: I'm not sure I like this model but I don't know enough there to reject it outright.

File: push_notifications/models.py [hunk 9]
```diff
@@ -74,7 +78,6 @@ def send_message(self, message, **kwargs):
 					if reg_ids:
 						r = gcm_send_message(reg_ids, data, cloud_type, application_id=app_id, **kwargs)
 						response.append(r)
-
```
Comment (jleclanche) [inline push_notifications/models.py]: No unnecessary changes in your commits please (keep style commits separate if you have to)

**Comment by goinnn** [comment]
I am going to update this pull request this week. :-)

File: setup.cfg (hunk 3 above)
Comment (goinnn) [reply setup.cfg]: Done [ca70992ea2f0e2b13d460c4adba5690bd4c2bc1e]

File: setup.cfg (hunk 6 above)
Comment (goinnn) [reply setup.cfg]: Done [5084ff16dd9c855cd087048b84eec9692592c228]

File: push_notifications/conf/app.py (hunk 2 above)
Comment (goinnn) [reply push_notifications/conf/app.py]: Done [19ac0c1e29402aee6a436726a3e892a08f2bc07f]

File: README.rst (hunk 4 above)
Comment (goinnn) [reply README.rst]: Done [40921fa09cc382653900d92497456afa46c2521c]

File: README.rst (hunk 5 above)
Comment (goinnn) [reply README.rst]: Done [40921fa09cc382653900d92497456afa46c2521c]

File: push_notifications/conf/app.py (hunk 2 above)
Comment (goinnn) [reply push_notifications/conf/app.py]: Now done [facad6cfc696a83c9667f194450b296c9e1f397f]

File: push_notifications/webpush.py (hunk 7 above)
Comment (goinnn) [reply push_notifications/webpush.py]: Same behaviour then apns2 [e443eb2b8d44b0f7e8079cec53b50afb10d1e365]

File: push_notifications/models.py (hunk 8 above)
Comment (goinnn) [reply push_notifications/models.py]: Improve it: [41e1fd3377cc5407d46c802797d3fadd1af861c9]

**Comment by goinnn** [comment]
I did every change that you asked me :-)

**Comment by goinnn** [comment]
@jleclanche @jamaalscarlett @matthewh is there any problem? If you need time to check it no problem, but I want to be sure that I did every change that you told me. Please could you confirm it?

**Comment by jamaalscarlett** [comment]
@goinnn I will take a look tonight EST

**Comment by goinnn** [comment]
@jamaalscarlett PING

**Comment by jamaalscarlett** [comment]
@goinnn Is there browser other than Firefox that implements the WebPush standard?  My understanding is that Chrome used GCM/FCM to receive notifications in the browser.

**Comment by jamaalscarlett** [comment]
Looking at the spec, only Firefox is supported.  I took a look at Google Chrome's push notification sample, and it works with Opera.  But as I said, that uses FCM.  At the very least, I would update the readme to explain that only Firefox is supported: https://developer.mozilla.org/en-US/docs/Web/API/Push_API#Browser_compatibility

**Comment by jamaalscarlett** [comment]
@goinnn Other than my comments about the readme file, it looks good to me.

**Comment by goinnn** [comment]
@jamaalscarlett I changed readme file with my changes in October. Please if I need any change more please tell me. And if it is right, please merge it :-)

**Comment by goinnn** [comment]
PING

**Comment by jamaalscarlett** [comment]
@goinnn Sorry, I have been swamped @ work, I will review this over the holiday.

**Comment by goinnn** [comment]
I'm sorry if I am a little annoying :-). I understand you are very busy. Me too :-). But I have to ping you another time :-)

**Comment by jleclanche** [comment]
Yeah.

**Comment by nlittlejohns** [comment]
Just wanted to say thanks for these changes, I think they fixed the error I was getting on Chrome.

Note to anyone who also gets this error:
```
Push failed: 403 Forbidden
Response body:aud claim in the token MUST include the origin of the push resource URL
```

Try updating to the latest code from git instead of running version 1.6.1 from PyPi.

Perhaps it's approaching time to make a new release?
